core/src/main/java/io/temporal/samples/hello/HelloWorkflowTimer.java

```diff
@@ -0,0 +1,253 @@
+import io.temporal.failure.ActivityFailure;
+import io.temporal.failure.CanceledFailure;
+import io.temporal.failure.ChildWorkflowFailure;
+import io.temporal.serviceclient.WorkflowServiceStubs;
+import io.temporal.worker.Worker;
+import io.temporal.worker.WorkerFactory;
+import io.temporal.workflow.*;
+import java.time.Duration;
+
+/** Sample shows how to use workflow timer instead of WorkflowOptions->Run/ExecutionTimeout */
+public class HelloWorkflowTimer {
+  private static String WORKFLOW_ID = "HelloWorkflowWithTimer";
+  private static String TASK_QUEUE = "HelloWorkflowWithTimerTaskQueue";
+  // Change time to 12 to 20 seconds to handle cancellation while child workflow is running
+  private static int TIME_SECS = 8;
+
+  // Workflow
+  @WorkflowInterface
+  public interface WorkflowWithTimer {
+    @WorkflowMethod
+    String execute(String input);
+  }
+
+  public static class WorkflowWithTimerImpl implements WorkflowWithTimer {
+    // Our timer cancellation scope
+    private CancellationScope timerCancellationScope;
+    // Our workflow cancellation scope
+    private CancellationScope workflowCancellationScope;
+    // Workflow result
+    private String workflowResult = "";
+    private Promise<Void> workflowTimerPromise;
+
+    @Override
+    public String execute(String input) {
+      // Create workflow timer (within timer cancel;ation scope so it can be canceled)
+      // which denotes the max amount of time we allow this execution to run
+      // Using workflow timer instead of workflow run/execution timeouts allow us to react to this
+      // timer
+      // fires, be able to chose if we want to fail or complete execution, and do some "cleanup"
+      // tasks if
+      // necessary before we do so. If we used workflow run/execution timeouts insted we would not
+      // be able
+      // to react to this timer firing (its server timer only)
+      timerCancellationScope =
+          Workflow.newCancellationScope(
+              () -> {
+                workflowTimerPromise =
+                    Workflow.newTimer(
+                            Duration.ofSeconds(TIME_SECS),
+                            TimerOptions.newBuilder().setSummary("Workflow Timer").build())
+                        // We can use thenApply here to cancel our cancelation scope when this timer
+                        // fires. Note we cannot complete the execution from here, see
+                        // https://github.com/temporalio/sdk-java/issues/87
+                        .thenApply(
+                            ignore -> {
+                              // Cancel the workflow cancellation scope allowing us to react to this
+                              // timer firing
+                              if (workflowCancellationScope != null) {
+                                workflowCancellationScope.cancel("Workflow timer fired");
+                              }
+                              return null;
+                            });
```

Comment:
I wonder if having this block inside `workflowCancellationScope` would produce the same output, so we might not need two cencellationScopes

Reply:
i think you could. having separate maybe cleaner as then we dont cancel scope within itself. wdyt?